security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,174 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/oauth"
+
+	mesh "istio.io/api/mesh/v1alpha1"
+	pb "istio.io/api/security/v1alpha1"
+	tokensource "istio.io/istio/pkg/util/tokensource"
+	"istio.io/istio/security/pkg/pki/signingapi"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultRequestTimeout = 10 * time.Second
+)
+
+var (
+	cLog = log.RegisterScope("CustomCAClient", "Custom CA Integration Log", 0)
+)
+
+// CAClient generates keys and certificates for Istio identities.
+type CAClient struct {
+	signingAPIClient *signingapi.Client
+	keyCertBundle    util.KeyCertBundle
+}
+
+// IsUsingCustomCAForWorkloadCertificate return true if users configure CustomCA for signing Workload Certificates
+// TODO: Support for Istio Agent side
+func IsUsingCustomCAForWorkloadCertificate(caOpts *mesh.MeshConfig_CA) bool {
+	return caOpts != nil && caOpts.Address != "" && caOpts.IstiodSide
+}
+
+// NewCAClient returns a new CAClient instance.
+// CustomCAClient will use JWT TOKEN from Istiod to create Authorization header each forwarding request to External CA
+// External CA just need to verify Istiod's identity each forwarding request is sufficient
+func NewCAClient(caOpts *mesh.MeshConfig_CA, keyCertBundle util.KeyCertBundle, jwtPath string) (*CAClient, error) {
+	requestTimeoutInSecond := defaultRequestTimeout
+
+	if caOpts.GetRequestTimeout() != nil {
+		requestTimeoutInSecond = time.Duration(caOpts.GetRequestTimeout().Seconds) * time.Second
+	}
+
+	signingAPIClient := signingapi.New(caOpts.GetAddress(), caOpts.TlsSettings, requestTimeoutInSecond)
+	if jwtPath != "" {
+		oauthTokenSource := oauth.TokenSource{TokenSource: &tokensource.FromFile{
+			Path:   jwtPath,
+			Period: time.Second * 300,
+		}}
+		signingAPIClient.AddDialOptions([]grpc.DialOption{grpc.WithPerRPCCredentials(oauthTokenSource)})
+	}
+
+	c := &CAClient{
+		keyCertBundle:    keyCertBundle,
+		signingAPIClient: signingAPIClient,
+	}
+
+	err := c.signingAPIClient.Connect(nil)
+	if err != nil {
+		cLog.Errorf("can not connect to Custom CA Address: %v", err)
+		return nil, fmt.Errorf("can not connect to Custom CA Address: %v", err)
+	}
+	return c, nil
+}
+
+func responseTimeTrack(start time.Time, name string) {
+	elapsed := time.Since(start)
+	cLog.Debugf("%s took %s", name, elapsed)
+}
+
+// CreateCertificate is similar to Sign but returns the leaf cert and the entire cert chain.
+func (c *CAClient) CreateCertificate(ctx context.Context,
+	req *pb.IstioCertificateRequest, identities []string) (*pb.IstioCertificateResponse, error) {
+	if err := verifyCSRIdentities(identities, req.GetCsr()); err != nil {
```

Comment:
It's here @rshriram ,
CSR's host is verified before forwarding to External CA.

Reply:
Are we also verifying that the incoming CSR doesn't also contain IPs, Common Name, and DNS?

Reply:
CSR from IstioAgent doesn't have IPs, Common Name, and DNS

Reply:
We should assume we receive malicious CSRs

Reply:
I'll add more code to make sure CSR doesn't contains IPs, Common Name, and DNS